public/app/features/plugins/components/AppRootPage.tsx

```diff
@@ -92,24 +93,33 @@ class AppRootPage extends Component<Props, State> {
   render() {
     const { loading, plugin, nav, portalNode } = this.state;
 
-    if (plugin && !plugin.root) {
-      // TODO? redirect to plugin page?
-      return <div>No Root App</div>;
+    if (!plugin) {
+      return <Page navId="apps">{loading && <PageLoader />}</Page>;
+    }
+
+    if (!plugin.root) {
+      <Page navId="apps">
+        <div>No root App</div>;
+      </Page>;
+    }
+
+    const pluginRoot = (
+      <plugin.root
+        meta={plugin.meta}
+        basename={this.props.match.url}
+        onNavChanged={this.onNavChanged}
+        query={this.props.queryParams as KeyValue}
+        path={this.props.location.pathname}
+      />
+    );
+
+    if (config.featureToggles.topnav && !nav) {
+      return pluginRoot;
```

Comment:
**Idea:** I was thinking about replacing `AppRootPage` with this [AppPluginLoader](https://github.com/grafana/grafana/blob/main/public/app/features/plugins/components/AppPluginLoader.tsx) component, and in there it would be probably easier to add these conditions than in here. (Or even using `AppPluginLoader` inside `AppRootPage` to make it simpler.)

Reply:
@leventebalogh Yea, using AppPluginLoader from AppRootPage looks possible, but the logic that exists here in this PR is needed for backward compatibility. that is AppRootPage wraps Plugin root in Page component always. the change here is that if topnav is enabled this is not done and we instead rely on plugins using PluginPage component from its root page